### PR TITLE
Various fix to handle issuer jwk correctly

### PR DESCRIPTION
**Pull request by PascalDR**

### CODE REVIEW

File: pyeudiw/openid4vp/__init__.py [hunk 1]
```diff
@@ -5,14 +5,16 @@
 from pyeudiw.sd_jwt import verify_sd_jwt
 
 
-def check_vp_token(vp_token: str, config: dict, sd_specification: dict, sd_jwt: dict) -> Tuple[str | None, dict]:
+def check_vp_token(vp_token: str, config: dict, sd_specification: dict, issuer_jwk: JWK, sd_jwt: dict = {"no_randomness": True}) -> Tuple[str | None, dict]:
```
Comment (peppelinux) [inline pyeudiw/openid4vp/__init__.py:8]: `sd_jwt: dict = {"no_randomness": True}`

please rename `sd_jwt` with something semantically more correct
I read functional parameters not an sd_jwt

File: pyeudiw/satosa/backend.py [hunk 2]
```diff
@@ -299,13 +292,16 @@ def redirect_endpoint(self, context, *args):
             if isinstance(decrypted_data["vp_token"], str)
             else decrypted_data["vp_token"]
         )
+        
+        # TODO: The issuer JWK must be selected from a dictinoary where the keys are the issuers and
+        # the values are the jwk connected to that issuer
+        issuer_jwk = JWK(self.config["metadata_jwks"][0])
```
Comment (peppelinux) [inline pyeudiw/satosa/backend.py:298]: this MUST be taken from the kid value not just the first value [0]

please use this:
https://github.com/italia/eudi-wallet-it-python/blob/dev/pyeudiw/federation/trust_chain_validator.py#L20

please move the previous function in pyeudiw.jwk.utils and rename that function to `find_jwk_by_kid`

File: pyeudiw/satosa/backend.py [hunk 3]
```diff
@@ -355,7 +351,16 @@ def redirect_endpoint(self, context, *args):
             all_user_claims, _info["issuer"]
         )
         
-        self.db_engine.update_response_object(nonce, state, internal_resp)
+        try:
+            self.db_engine.update_response_object(nonce, state, internal_resp)
+        except Exception as e:
+            return JsonResponse(
```
Comment (peppelinux) [inline pyeudiw/satosa/backend.py:357]: please add the logging call here

File: pyeudiw/satosa/backend.py [hunk 4]
```diff
@@ -424,10 +429,19 @@ def request_endpoint(self, context, *args):
...
-            context.http_headers['HTTP_DPOP'], 
-            context.http_headers['HTTP_AUTHORIZATION']
-        )
+        try:
+            entity_id = self.db_engine.init_session(
+                dpop_proof=context.http_headers['HTTP_DPOP'], 
+                attestation=context.http_headers['HTTP_AUTHORIZATION']
+            )
+        except Exception as e:
+            return JsonResponse(
```
Comment (peppelinux) [inline pyeudiw/satosa/backend.py:438]: please add the logging call here

File: pyeudiw/satosa/backend.py [hunk 5]
```diff
@@ -450,7 +464,16 @@ def request_endpoint(self, context, *args):
         jwt = helper.sign(data)
         response = {"response": jwt}
         
-        self.db_engine.update_request_object(entity_id, nonce, state, data)
+        try:
+            self.db_engine.update_request_object(entity_id, nonce, state, data)
+        except Exception as e:
+            return JsonResponse(
```
Comment (peppelinux) [inline pyeudiw/satosa/backend.py:470]: please add the logging call here

**Review by peppelinux — CHANGES_REQUESTED** [review]

File: pyeudiw/satosa/backend.py (hunk 3 above)
Comment (PascalDR) [reply pyeudiw/satosa/backend.py:357]: Resolved in [c69e7db](https://github.com/italia/eudi-wallet-it-python/pull/89/commits/c69e7db13eddb6b3a16fc2e5d182fea4b08cfba8)

File: pyeudiw/satosa/backend.py (hunk 4 above)
Comment (PascalDR) [reply pyeudiw/satosa/backend.py:439]: Resolved in [c69e7db](https://github.com/italia/eudi-wallet-it-python/pull/89/commits/c69e7db13eddb6b3a16fc2e5d182fea4b08cfba8)

File: pyeudiw/satosa/backend.py (hunk 5 above)
Comment (PascalDR) [reply pyeudiw/satosa/backend.py:472]: Resolved in [c69e7db](https://github.com/italia/eudi-wallet-it-python/pull/89/commits/c69e7db13eddb6b3a16fc2e5d182fea4b08cfba8)

File: pyeudiw/satosa/backend.py (hunk 2 above)
Comment (PascalDR) [reply pyeudiw/satosa/backend.py:297]: The kid is the one present in the vp_token header?

Comment (peppelinux) [reply pyeudiw/satosa/backend.py:297]: yes

Comment (PascalDR) [reply pyeudiw/satosa/backend.py]: Resolved in [adfc12f](https://github.com/italia/eudi-wallet-it-python/pull/89/commits/adfc12fd4c6933fca8ea0c9d2655712c46970107)

File: pyeudiw/satosa/backend.py [hunk 6]
```diff
@@ -230,20 +229,18 @@ def _translate_response(self, response, issuer):
...
 
     def _handle_vp(self, vp_token: str, context: Context) -> dict:
-        valid, value = check_vp_token(
-            vp_token, self.config, self.sd_specification, self.sd_jwt)
+        valid, value = None, None
+        
+        try:
+            valid, value = check_vp_token(
+                vp_token, None, self.metadata_jwks_by_kids)
+        except Exception as e:
```
Comment (peppelinux) [inline pyeudiw/satosa/backend.py:237]: here the exception should be specialized for this case, using the already defined exception InvalidVPToken

**Review by peppelinux — CHANGES_REQUESTED** [review]

File: pyeudiw/openid4vp/__init__.py [hunk 7]
```diff
@@ -1,18 +1,28 @@
...
-def check_vp_token(vp_token: str, config: dict, sd_specification: dict, sd_jwt: dict) -> Tuple[str | None, dict]:
+def check_vp_token(vp_token: str, sd_specification: dict, jwks: list[dict], config: dict = {"no_randomness": True}) -> Tuple[str | None, dict]:
     payload = unpad_jwt_payload(vp_token)
-    holder_jwk = JWK(payload["cnf"]["jwk"])
-    issuer_jwk = JWK(config["federation"]["federation_jwks"][1])
-
-    result, binding = verify_sd_jwt(
-        vp_token, sd_specification, sd_jwt, issuer_jwk, holder_jwk)
-    nonce = binding.get("nonce", None)
+    kid = unpad_jwt_header(vp_token)["kid"]
```
[the remark this answers is not in the document]
Comment (PascalDR) [reply pyeudiw/openid4vp/__init__.py]: Resolved in commit [e4ec122](https://github.com/italia/eudi-wallet-it-python/pull/89/commits/e4ec122051f5f18b638bb62ab2f347733249cab6)

File: pyeudiw/satosa/backend.py (hunk 6 above)
Comment (PascalDR) [reply pyeudiw/satosa/backend.py]: Resolved in [6f6ce24](https://github.com/italia/eudi-wallet-it-python/pull/89/commits/6f6ce24fe06dbf4c63d54a8ae9e2b2090a8f65f5)

File: pyeudiw/satosa/backend.py [hunk 8]
```diff
@@ -230,20 +229,18 @@ def _translate_response(self, response, issuer):
...
-            _msg = "vp_token's nonce not present"
-            self._log(context, level='error', message=_msg)
-            return JsonResponse(
-                {
-                    "error": "parameter_absent",
-                    "error_description": _msg
-                },
-                status="400"
-            )
+            raise NoNonceInVPToken("vp_token's nonce not present")
```
[the remark this answers is not in the document]
Comment (PascalDR) [reply pyeudiw/satosa/backend.py]: Resolved in [6f6ce24](https://github.com/italia/eudi-wallet-it-python/pull/89/commits/6f6ce24fe06dbf4c63d54a8ae9e2b2090a8f65f5)

File: pyeudiw/satosa/backend.py [hunk 9]
```diff
@@ -439,6 +443,7 @@ def request_endpoint(self, context, *args):
                 attestation=context.http_headers['HTTP_AUTHORIZATION']
             )
         except Exception as e:
+            self._log(context, level='error', message=str(e))
```
Comment (peppelinux) [inline pyeudiw/satosa/backend.py:446]: str(e) is not enough, please envelope it in a format string and give more context about the nature of the exception

**Review by peppelinux — CHANGES_REQUESTED** [review]

File: pyeudiw/satosa/backend.py (hunk 9 above)
Comment (PascalDR) [reply pyeudiw/satosa/backend.py:424]: Resolved in [ec8559a](https://github.com/italia/eudi-wallet-it-python/pull/89/commits/ec8559aa058253962e9d78bae647587522324e63)

**Review by peppelinux — APPROVED** [review]

**Pull request merged**